processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java

```diff
@@ -467,6 +449,45 @@ else if ( collectionOrMap.isMapType()
         return collectionOrMap;
     }
 
+    /**
+     * Establishes the type bound:
+     * <ol>
+     * <li>{@code<? extends Number>}, returns Number</li>
+     * <li>{@code<? super Number>}, returns Number</li>
+     * <li>{@code<?>}, returns Object</li>
+     * <li>{@code<T extends Number>, returns Number}</li>
+     * </ol>
+     *
+     * @param typeMirror the type to return the bound for
+     * @return the bound for this parameter
+     */
+    public TypeMirror getTypeBound(TypeMirror typeMirror) {
```

Comment:
I only moved that logic out from `Types` and the corresponding method there was `getTypeBound` as well. I think as it also extracts a lower bound (from `? super X`)... so...
But as Sjaak wrote in the other PR, our handling of type-var bounds might be a bit rough anyway and it's something we need to dig in deeper.